### PR TITLE
For start end step

### PR DESCRIPTION
**Pull request by JessonChan**
close https://github.com/goplus/gop/issues/875

### CODE REVIEW

File: cl/stmt.go [hunk 1]
```diff
@@ -333,6 +341,44 @@ func compileForPhraseStmt(ctx *blockCtx, v *ast.ForPhraseStmt) {
...
+	}
+	if re.Expr3 == nil {
+		re.Expr3 = &ast.BasicLit{ValuePos: forPos, Kind: token.INT, Value: "1"}
+	}
+	return &ast.ForStmt{
+		For: forPos,
+		Init: &ast.AssignStmt{
+			Lhs:    []ast.Expr{value},
+			TokPos: re.To,
+			Tok:    token.DEFINE,
```
Comment (xushiwei) [inline cl/stmt.go:359]: maybe it's a `assign` statement. eg. 

```go
for x = range 1:6:2 {
}
```

File: cl/stmt.go [hunk 2]
```diff
@@ -333,6 +341,44 @@ func compileForPhraseStmt(ctx *blockCtx, v *ast.ForPhraseStmt) {
...
+			Rhs:    []ast.Expr{re.First},
+		},
+		Cond: &ast.BinaryExpr{
+			X:     value,
+			OpPos: re.To,
+			Op:    token.LSS,
+			Y:     re.Last,
```
Comment (xushiwei) [inline cl/stmt.go:366]: maybe `re.Last` is a `function call`. it's not reasonable to call `re.Last` multiple times.

File: cl/stmt.go [hunk 3]
```diff
@@ -333,6 +341,44 @@ func compileForPhraseStmt(ctx *blockCtx, v *ast.ForPhraseStmt) {
...
+		},
+		Post: &ast.AssignStmt{
+			Lhs:    []ast.Expr{value},
+			TokPos: re.Colon2,
+			Tok:    token.ASSIGN,
+			Rhs: []ast.Expr{&ast.BinaryExpr{
+				X:  value,
+				Op: token.ADD,
```
Comment (xushiwei) [inline cl/stmt.go:374]: why not use `ADD_ASSIGN`?

File: cl/stmt.go [hunk 4]
```diff
@@ -333,6 +341,44 @@ func compileForPhraseStmt(ctx *blockCtx, v *ast.ForPhraseStmt) {
...
+				Y:  re.Expr3,
```
Comment (xushiwei) [inline cl/stmt.go:375]: maybe `re.Expr3` is a `function call`. it's not reasonable to call `re.Expr3` multiple times.

File: cl/compile_test.go [hunk 5]
```diff
@@ -3355,19 +3334,97 @@ for range :10 {
...
+	func (t T)end()int{
+		return 3 
+	}
+	func (t T)step()int{
+		return 1 
+	}
+	
+	t:=T{}
+	
+	for i <- t.start():t.end():t.step(){
```
Comment (xushiwei) [inline cl/compile_test.go:3405]: missing a testcase:

```for
var i int
for i = range t.start():t.end():t.step() {
}
```

File: cl/stmt.go [hunk 6]
```diff
@@ -260,6 +260,14 @@ func compileAssignStmt(ctx *blockCtx, expr *ast.AssignStmt) {
 //    body
 // end
 func compileRangeStmt(ctx *blockCtx, v *ast.RangeStmt) {
+	if re, ok := v.X.(*ast.RangeExpr); ok {
+		tok := token.DEFINE
+		if v.Tok == token.ASSIGN {
```
Comment (xushiwei) [inline cl/stmt.go:265]: tok := v.Tok is not ok?

Comment (JessonChan) [reply cl/stmt.go:265]: there is no tok in this block
```go
for range [1,2,3]{
     println
}
```

**Pull request merged**